tracing-subscriber/src/registry/mod.rs

```diff
@@ -45,3 +42,188 @@ pub trait LookupMetadata {
         self.metadata(id).is_some()
     }
 }
+
+/// Provides access to stored span data.
+///
+/// Subscribers which store span data and associate it with span IDs should
+/// implement this trait; if they do, any [`Layer`]s wrapping them can look up
+/// metadata via the [`Context`] type's [`span()`] method.
+///
+/// [`Layer`]: ../layer/struct.Layer.html
+/// [`Context`]: ../layer/struct.Context.html
+/// [`span()`]: ../layer/struct.Context.html#method.metadata
+pub trait LookupSpan<'a> {
+    /// The type of span data stored in this registry.
+    type Data: SpanData<'a>;
+
+    /// Returns the [`SpanData`] for a given `Id`, if it exists.
+    ///
+    /// **Note**: users of the `LookupSpan` trait should typically call the
+    /// [`span`] method rather than this method. The `span` method is
+    /// implemented by calling `span_data`, but returns a reference which is
+    /// capable of performing more sophisiticated queries.
+    ///
+    /// [`SpanData`]: trait.SpanData.html
+    /// [`span`]: #method.span
+    fn span_data(&'a self, id: &Id) -> Option<Self::Data>;
+
+    /// Returns a [`SpanRef`] for the span with the given `Id`, if it exists.
+    ///
+    /// A `SpanRef` is similar to [`SpanData`], but it allows performing
+    /// additional lookups against the registryr that stores the wrapped data.
+    ///
+    /// In general, _users_ of the `LookupSpan` trait should use this method
+    /// rather than the [`span_data`] method; while _implementors_ of this trait
+    /// should only implement `span_data`.
+    ///
+    /// [`SpanRef`]: struct.SpanRef.html
+    /// [`SpanData`]: trait.SpanData.html
+    /// [`span_data`]: #method.span_data
+    fn span(&'a self, id: &Id) -> Option<SpanRef<'_, Self>>
+    where
+        Self: Sized,
+    {
+        let data = self.span_data(id)?;
+        Some(SpanRef {
+            registry: self,
+            data,
+        })
+    }
+}
+
+/// A stored representation of data associated with a span.
+pub trait SpanData<'a> {
+    /// An iterator of all the spans this span succeeds.
+    type Follows: Iterator<Item = &'a Id>;
```

Comment:
we should remove this from the API for now, since it isn't currently implemented.